File: frontend/preview/server/resources.js
/*
 *   Copyright (c) 2024 Jack Bennett.
 *   All Rights Reserved.
 *
 *   See the LICENCE file for more information.
 */

import fs from "fs/promises";
import path from "path";

var cachedResources = {};

var loadError;

const basePath = path.join(import.meta.dirname, "..");

export async function loadResource(filename, cache = true) {
    var cont;

    if (cache && (filename in cachedResources)) {
        return cachedResources[filename];
    }

    try {
        cont = (await fs.readFile(`${basePath}/${filename}`));
    } catch (err) {
        loadError = err;
        return undefined;
    }

    if (cache) {
        cachedResources[filename] = cont;
    }

    return cont;
}

export function getLastLoadError() {
    return loadError;
}

File: texedit/util/resources.cpp
/*
 *   Copyright (c) 2024 Jack Bennett.
 *   All Rights Reserved.
 *
 *   See the LICENCE file for more information.
 */

#include "resources.hpp"

#include <wx/filename.h>
#include <wx/dir.h>
#include <wx/stdpaths.h>

namespace te::util {
    wxString GetTexeditDir() {
        wxFileName f = wxStandardPaths::Get().GetExecutablePath();
        return f.GetPath();
    }

    wxString RelToExec(const std::vector<wxString> &dirs, const wxString &filename) {
        wxFileName f = wxFileName::DirName(GetTexeditDir());
        for (auto d : dirs) {
            f.AppendDir(d);
        }

        return f.GetPath() +
               wxFileName::GetPathSeparator() +
               filename;
    }

    wxString RelToExec(const wxString &filename) {
        return RelToExec({}, filename);
    }

    bool ValidateExecutable(const wxString &path) {
        wxFileName f{path};
        return f.FileExists() && f.IsFileExecutable();
    }
}
